audbackend/core/artifactory.py

```diff
@@ -154,6 +154,38 @@ def __init__(
         )
         self._repo = path.find_repository_local(self.repository)
 
+        self.extensions = []
```

Comment:
If we think this might be needed by other backend implementations, too. We could also move it to `Backend` and leave it to the backend implementation if they use the information or not.

Reply:
The default value is an empty list and in the docstring we show an example how to add an extension to it, so I guess it should be fine.

Reply:
If we decide to make it available to all backends we could make `Backend.extensions` a property that is set with `create()` and `access()`. The nicest solution would be to have it only with `create()`, but backends that use `extensions` must store them somehow with the repository (e.g. in a hidden file in the root directory).

Reply:
If I understand it correctly (compare https://github.com/audeering/audbackend/pull/130#issuecomment-1547269016), we need `Artifactort.extensions` only to handle already existing repositories that have used extensions including a dot.

In order to allow backward compatibility for all backends it would indeed make sense to provide it not only for `Artifactory`, but for `Backend`. On the other hand, this only makes sense when also switching back to the old file structure on the `file-system` backend.

As a feature for new repositories I found `extensions` a little bit risky as now a user will not only have to know the backend and repository name, but also the `Backend.extensions` settings needed to work with an existing repository. So it looks like this property would then need to be stored in the repository itself. But I would not go that far. Instead I would vote to rename it to `Backend.legacy_extensions()` and not include it in the default API documentation.
Instead I would add a section to the documentation discussing how to handle already existing repositories with the new `audbackend` version. This documentation, I would propose to add in another pull request.

Reply:
>> On the other hand, this only makes sense when also switching back to the old file structure on the file-system backend.
>
> I think it's ok if we are not backward compatible with file-system. So I would not switch back to the old structure.

I agree, and created https://github.com/audeering/audbackend/issues/131 to track documenting this.

> Ok, but then I propose we also use `<host>/<repository>/<folder>/<name-wo-ext>/<version>/<name-wo-ext>-<version>.<ext>` only on legacy repositories and otherwise switch to `<host>/<repository>/<folder>/<version>/<name>.<ext>`. Without an official way to specify custom extensions the old structure makes no sense to me since e.g. file.tar.gz will be stored as `'.../file.tar/1.0.0/file.tar-1.0.0.gz'`.

I disagree here as it again would make things more complicated then is needed.
I would always stick to `<host>/<repository>/<folder>/<name-wo-ext>/<version>/<name-wo-ext>-<version>.<ext>`,
and for the new Artifactory backend I would not provide a way to configure what is a valid file extension. This way the resulting path on the backend will always be easily predictable and not depend on a user setting.
I don't see a big issue in having `'.../file.tar/1.0.0/file.tar-1.0.0.gz'` as path.

Reply:
> I don't see a big issue in having '.../file.tar/1.0.0/file.tar-1.0.0.gz' as path.

I think it can be an issue. E.g. if they provide a way to preview TAR.GZ files on Artifactory as they do with ZIP in their UI. And also generally, it doesn't feel right to use `gz` as file extension when the extension is `tar.gz`.

Reply:
And also don't forget that one reason we added the `ext` argument was to group files in the same directory on Artifactory, e.g. the YAML files that store header and the meta information in `audmodel`.

Reply:
The only save usage of letting the user configure possible additional extension names seems to me during the creation of the repository. After the first files have been uploaded the user shouldn't be allowed to change the file extension settings as this might lead to the case that certain files can no longer be accessed.

Reply:
> The only save usage of letting the user configure possible additional extension names seems to me during the creation of the repository.

Yes, and that's a strong argument we should not generally switch back to the old file structure as it only makes things complicated. Instead we keep the new file structure and only introduce a legacy function that allows to switch to the old file structure in packages where relied on it on the past.

Reply:
I guess we are stuck here again ;)

My requirement would still be that all old and new versions of `audb` should work with already existing repos without that the user needs to do anything. And so far I did not see how this will work when we switch to another file structure on the backend.

Reply:
But with `audb` it will use the old structure, even when we create new repositories since we hard code this on package level. Only new packages, where we don't force the legacy package will use the structure. This makes since we don't want the user to bother file extensions.

Reply:
Ah, ok, now I get it. So basically with `Artifactory._use_legacy_file_structure(extensions=[])` we can solve all the issues and it also makes sense to implement both inside this pull request.

We could add the switch-function also to the `file-system` backend then we don't need to add extra explanation how to convert existing `file-system` backends and could also continue in `audb` using the old structure.

Reply:
I added:

![image](https://github.com/audeering/audbackend/assets/10383417/1f21820f-896b-48ba-9822-3051c86de0dd)

and updated the description accordingly.

Reply:
In https://github.com/audeering/audbackend/pull/130#discussion_r1193651913 I suggested to add the legacy function also to the `FileSystem` backend as the user could then mimic how the files would be stored on the Artifactory server. But as we want to support several backends in the long run and each might store the files differently, there might not be a need to add the legacy method to `FileSystem` as well (I would anyway not do this inside this pull request).

On the other hand, if a University have used `audb` together with a NAS and the `FileSystem` backend it would still ensure backward compatibility in that case for `audb`.

What would you prefer?

Reply:
I don't see an advantage if we mimic the file system on the `Artifactory` with `FileSystem`. 

> On the other hand, if a University has used audb together with a NAS and the FileSystem backend it would still ensure backward compatibility in that case for audb.

What are the chances? :) But if we want to backward compatible, we could apply the same solution to `FileSystem`.

Reply:
I'm also not completely decided, but I created https://github.com/audeering/audbackend/issues/136 to track this.